src/sysupdate/sysupdate-resource.c

```diff
@@ -596,12 +633,22 @@ int resource_resolve_path(
                         return log_error_errno(SYNTHETIC_ERRNO(EPERM),
                                                "Block device is not allowed when using --root= mode.");
 
-                r = stat("/run/systemd/volatile-root", &orig_root_stats);
+                struct stat orig_root_stats;
+                r = RET_NERRNO(stat("/run/systemd/volatile-root", &orig_root_stats));
                 if (r < 0) {
-                        if (errno == ENOENT) /* volatile-root not found */
-                                r = get_block_device_harder("/usr/", &d);
-                        else
+                        if (r != -ENOENT)
                                 return log_error_errno(r, "Failed to stat /run/systemd/volatile-root: %m");
+
+                        /* volatile-root not found */
+                        r = get_block_device_harder("/usr/", &d);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to determine block device of file system: %m");
+                        if (r == 0) {
+                                /* Not backed by a block device? Let's see if this is a sysext overlayfs instance */
+                                r = get_sysext_overlay_block("/usr/", &d);
+                                if (r < 0)
+                                        return log_error_errno(r, "Failed to determine block device of file system: %m");
```

Comment:
Not a matter, but
```c
r = get_block_device_harder(...);
if (r == 0)
        r = get_sysext_overlay_block(...);
if (r < 0)
        return log_error_errno(...);
```
is shorter if log messages for the two failures are equivalent.